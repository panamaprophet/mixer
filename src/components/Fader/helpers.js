'use strict';


import {min, max} from 'ramda';


/**
 * @typedef {Object} PointerPosition
 * @property {number} value - absolute value in pixels
 * @property {number} percentage - relative value in percents
 */


/**
 * @returns {number} vertical coordinate from touch or mouse event
 */
export const getY = event => event.touches ? event.touches[0].pageY : event.pageY;

/**
 * @returns {number} horizontal coordinate from touch or move event
 */
export const getX = event => event.touches ? event.touches[0].pageX : event.pageX;

/**
 * @returns {PointerPosition}
 */
export const getPointerVerticalPosition = (position, {top, bottom, height}) => {
    const value = bottom - min(min(position, top), bottom);

    return {
        value,
        percentage: value / (height / 100),
    };
};

/**
 * @returns {PointerPosition}
 */
export const getPointerHorizontalPosition = (position, {width, left}) => {
    const value = max(min(position, left + width), left) - left;

    return {
        value,
        percentage: value / (width / 100),
    }
};