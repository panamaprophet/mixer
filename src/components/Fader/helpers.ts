'use strict';


import {min, max} from 'ramda';

/**
 * @returns {number} vertical coordinate from touch or mouse event
 */
export const getY = event => event.touches ? event.touches[0].pageY : event.pageY;

/**
 * @returns {number} horizontal coordinate from touch or move event
 */
export const getX = event => event.touches ? event.touches[0].pageX : event.pageX;

/**
 * @returns {number} relative position in percents
 */
export const getPointerVerticalPosition = (position, {top, bottom, height}) => {
    let value = position;

    if (value < top) value = top;
    if (value > bottom) value = bottom;

    value = bottom - value;

    return value / (height / 100);
};

/**
 * @returns {number} relative position in percents
 */
export const getPointerHorizontalPosition = (position, {width, left}) => {
    let value = position; // max(min(position, left + width), left) - left;

    const rightBorder = left + width;

    if (value > rightBorder) value = rightBorder;
    if (value < left) value = left;

    value = value - left;

    return value / (width / 100);
};