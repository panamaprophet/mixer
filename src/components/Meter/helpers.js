import {converge, divide, sum, length} from 'ramda';


const COLOR_GREEN = '#05860f';
const COLOR_YELLOW = '#f9ef5c';
const COLOR_RED = '#861615';

/**
 * @param {AudioContext} context
 * @param {Canvas} canvas
 * @returns {CanvasGradient}
 */
export const createMeterGradient = (context, {
    width,
    height,
}) => {
    const gradient = context.createLinearGradient(0, 0, width, height);

    gradient.addColorStop(0.0, COLOR_GREEN);
    gradient.addColorStop(0.8, COLOR_YELLOW);
    gradient.addColorStop(0.9, COLOR_RED);

    return gradient;
}

export const getAverage = converge(divide, [sum, length]);