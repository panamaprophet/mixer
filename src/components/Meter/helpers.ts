import {converge, divide, sum, length} from 'ramda';


const LEVEL_LOW = '#05860f';

const LEVEL_MIDDLE = '#f9ef5c';

const LEVEL_HIGH = '#861615';


/**
 * @param {AudioContext} context
 * @param {Canvas} canvas
 * @returns {CanvasGradient}
 */
export const createMeterGradient = (context: CanvasRenderingContext2D, {
    width,
    height,
}: {
    width: number,
    height: number,
}): CanvasGradient => {
    const gradient = context.createLinearGradient(0, 0, width, height);

    gradient.addColorStop(0.0, LEVEL_LOW);
    gradient.addColorStop(0.8, LEVEL_MIDDLE);
    gradient.addColorStop(0.9, LEVEL_HIGH);

    return gradient;
}

export const getAverage = converge(divide, [sum, length]);