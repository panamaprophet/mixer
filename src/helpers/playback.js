import {map} from 'ramda';

import {
	PLAYBACK_STATUS,
} from '/constants';


export const isPlaying = controls => controls.status === PLAYBACK_STATUS.PLAYING;

export const isPaused = controls => controls.status === PLAYBACK_STATUS.PAUSED && controls.currentPosition !== 0;

export const playAll = map(track => track.play());

export const pauseAll = map(track => track.pause());

export const rewindAll = map(track => {
    track.stop();
    track.play();
});