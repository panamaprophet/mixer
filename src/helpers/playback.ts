import {map, not} from 'ramda';

import {
    PLAYBACK_STATUS,
} from '/constants';


type Track = {
    play: () => void,
    pause: () => void,
    stop: () => void,
};


export const isPlaying = playback => playback.status === PLAYBACK_STATUS.PLAYING;

export const isPaused = playback => playback.status === PLAYBACK_STATUS.PAUSED; //@TODO: && playback.currentPosition !== 0;

export const isReady = playback => playback.status === PLAYBACK_STATUS.READY;

export const isActive = playback => (isPlaying(playback) || isPaused(playback) || isReady(playback));

export const isNotActive = playback => not(isActive(playback));

export const playAll = map((track: Track) => track.play());

export const pauseAll = map((track: Track) => track.pause());

export const rewindAll = map((track: Track) => {
    track.stop();
    track.play();
});

export const stopAll = map((track: Track) => track.stop());