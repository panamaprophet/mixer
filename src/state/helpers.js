import {
    getNodeParamNormalizedValue,
} from '/helpers/node';


export const STATUS = {
    'PLAYING': 'PLAYING',
    'PAUSED': 'PAUSED',
};


const createTrackEntity = ({id, title, volume, muted, bypassFX, playing, fx}) => ({
    id,
    title,
    volume,
    isMuted: muted,
    isEffectsDisabled: bypassFX,
    isPlaying: playing,
    send: {
        delay: getNodeParamNormalizedValue(fx.delay.gain),
        reverb: getNodeParamNormalizedValue(fx.reverb.gain),
        distortion: getNodeParamNormalizedValue(fx.distortion.gain),
    },
});

const createEffectEntity = ({id, ...rest}) => ({
    id,
    rest,
});


export const createState = mixdesk => {
    const tracks = mixdesk.tracks.map(createTrackEntity);
    const effects = mixdesk.fx.map(createEffectEntity);
    const analyser = mixdesk.analyser;

    const controls = {
        analyser,
        status: STATUS.PAUSED,
    };

    return {
        tracks,
        effects,
        controls,
    };
};