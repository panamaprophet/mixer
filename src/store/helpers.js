import React, {useEffect} from 'react';
import {map, curry, reject, not} from 'ramda';

import {
    TRACK_STATE,
} from '/constants';

import {
    createEffectEntity,
    createPlaybackEntity,
    createTrackEntity,
} from '/helpers/entities';

import {mixdesk} from './mixdesk';


const compact = reject(item => not(Boolean(item)));

const getLoadingState = track => track.loadingState;

const getLoadingStates = ({tracks}) => map(getLoadingState, tracks);

/**
 * Dispatches set ready state action with provided track params
 *
 * @param {function} dispatch
 * @param {Track}
 */
const dispatchSetTrackState = curry((dispatch, {id, state}) => dispatch({
    type: 'TRACK_SET_READY_STATE',
    payload: {
        trackId: id,
        state,
    },
}));

/**
 * Updates tracks statuses in store
 *
 * @param {function} — dispatch
 * @returns {Promise<Track[]>}
 */
export const setTracksStatusesOnLoad = dispatch => 
    Promise
        .all(getLoadingStates(mixdesk))
        .then(map(dispatchSetTrackState(dispatch)));

/**
 * Brings dispatch arguments to console
 *
 * @param {function} — original dispatch
 * @returns {function} — curried version with original dispatch binded
 */
export const getDispatchWithLog = curry((dispatch, args) => {
    console.log('[DISPATCH LOG]', args);

    return dispatch(args);
});

export const createState = mixdesk => {
    const tracks = compact(mixdesk.tracks.map(createTrackEntity));
    const effects = compact(mixdesk.fx.map(createEffectEntity));
    const playback = createPlaybackEntity({
        analyser: mixdesk.analyser,
    });

    return {
        tracks,
        effects,
        playback,
    };
}