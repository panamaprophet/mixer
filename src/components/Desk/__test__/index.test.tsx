import React from 'react';
import {shallow} from 'enzyme';
import {Desk} from '../';
import {PlaybackStatus} from '/helpers/playback';


const playbackMock = {
    status: PlaybackStatus.PAUSED,
    currentPosition: 0,
};

const props = {
    onPlay: jest.fn(),
    onPause: jest.fn(),
    onRewind: jest.fn(),
    playback: playbackMock,
}

const PLAY_BUTTON = 0;
const PAUSE_BUTTON = 1;
const REWIND_BUTTON = 2;


describe('<Desk />', () => {
    it('should renders without any errors', () => {
        const wrapper = shallow(<Desk {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    describe('playback buttons', () => {
        it('should be disabled if mixdesk is not active', () => {
            const p = {
                ...props,
                playback: {
                    ...props.playback,
                    status: PlaybackStatus.NOT_SET,
                },
            };

            const wrapper = shallow(<Desk {...p} />);
            const result = wrapper.find('button[disabled]').length;

            expect(result).toBe(3);
        });
    });

    describe('play button', () => {
        it('should call onPlay on press', () => {
            const wrapper = shallow(<Desk {...props} />);

            wrapper.find('button').at(PLAY_BUTTON).simulate('click');

            expect(props.onPlay).toBeCalled();
        });

        it('should have active class on playing', () => {
            const p = {
                ...props,
                playback: {
                    ...props.playback,
                    status: PlaybackStatus.PLAYING,
                },
            };

            const wrapper = shallow(<Desk {...p} />);
            const result = wrapper.find('button').at(PLAY_BUTTON).hasClass('isButtonPressed');

            expect(result).toBe(true);
        });
    });

    describe('pause button', () => {
        it('should call onPause on press', () => {
            const wrapper = shallow(<Desk {...props} />);
            wrapper.find('button').at(PAUSE_BUTTON).simulate('click');

            expect(props.onPause).toBeCalled();
        });

        it('should have active class on pause', () => {
            const p = {
                ...props,
                playback: {
                    ...props.playback,
                    status: PlaybackStatus.PAUSED,
                    currentPosition: 420,
                },
            };

            const wrapper = shallow(<Desk {...p} />);
            const result = wrapper.find('button').at(PAUSE_BUTTON).hasClass('isButtonPressed');

            expect(result).toBe(true);
        });
    });

    describe('rewind button', () => {
        it('should call onRewind on press', () => {
            const wrapper = shallow(<Desk {...props} />);

            wrapper.find('button').at(REWIND_BUTTON).simulate('click');

            expect(props.onRewind).toBeCalled();
        });
    });
});