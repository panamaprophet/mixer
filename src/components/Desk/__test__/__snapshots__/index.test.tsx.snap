// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Desk /> should renders without any errors 1`] = `
<div
  className="desk"
>
  <div
    className="tracks"
  />
  <div
    className="controlsContainer"
  >
    <div
      className="controls"
    >
      <button
        className="control button"
        disabled={false}
        onClick={[MockFunction]}
      >
        <Icon
          type="play"
        />
        Play
      </button>
      <button
        className="control button isButtonPressed"
        disabled={false}
        onClick={[MockFunction]}
      >
        <Icon
          type="pause"
        />
        Pause
      </button>
      <button
        className="control button"
        disabled={false}
        onClick={[MockFunction]}
      >
        <Icon
          type="rewind"
        />
        Rewind
      </button>
    </div>
    <div
      className="sends"
    />
  </div>
</div>
`;
